### PR TITLE
Add SignalSnakeJs

### PR DESCRIPTION
**Pull request by milesfrain**
Demo:
https://try.ps.ai/?github=milesfrain/purescript-cookbook/signal-snake-js/recipes/SignalSnakeJs/src/Main.purs

If this example isn't too confusing for beginners, it could replace #212.

### CODE REVIEW

**Comment by milesfrain** [comment]
Keeping this as a draft to give @holdenlee a chance to share citation preferences.
(Discussed in https://github.com/holdenlee/purescript-games/issues/4)

File: recipes/SignalSnakeJs/src/Main.purs [hunk 1]
```diff
@@ -0,0 +1,369 @@
...
+-- is used here to provide a signal value at time = 0.
+mapKey :: Direction -> Signal Boolean -> Signal Direction
+mapKey dir sig = filterMap (fromBool dir) initialDirection sig
+
+-- Helper function for mapKey's filterMap.
+-- Convert's a Boolean to a Maybe using a default value.
+fromBool :: forall a. a -> Boolean -> Maybe a
+fromBool x b
+  | b = Just x
+  | otherwise = Nothing
```
Comment (milesfrain) [inline recipes/SignalSnakeJs/src/Main.purs:275]: I suspected something like this would already exist, but I couldn't find anything on Pursuit.

**Comment by JordanMartinez** [comment]
I don't believe we should merge this until a license for that work is issued. I can't find any on the repo.

That's the only blocker in my mind.

**Comment by holdenlee** [comment]
I give permission to use the example in the PureScript Cookbook.

**Comment by afcondon** [comment]
One style suggestion (and it could well be that you disagree with it, but i'll just put it here for consideration) is that you could use a case instead of the nested if-then-else, ie

```    
    case willDie, willEat of
         false  , false -> m { snake = moveSnakeNoGrow nextPoint m.snake }
         true   , _     -> m { snake = initialSnake, direction = initialDirection }
         false  , true  -> 
          let 
            biggerSnake = moveSnakeAndGrow nextPoint m.snake
            food /\ genState = runGen (availableRandomPoint biggerSnake) m.genState
          in
            m
              { snake = biggerSnake
              , food = food
              , genState = genState
              }
```

**Comment by milesfrain** [comment]
> use a case instead of the nested if-then-else

Good suggestion. Now if only [purty would let me keep those alignment spaces](https://gitlab.com/joneshf/purty/-/issues/200)...

**Comment by afcondon** [comment]
Yes, i love the existence of, and many ideas in, Purty, but to me compromising readability of the actual code in order to maximize readability of commits is putting the cart before the horse.

Very possibly not worth doing if formatting messes it up, if-then-else isn't the worst thing in the world.

File: recipes/SignalSnakeJs/src/Main.purs (hunk 1 above)
Comment (paluh) [reply recipes/SignalSnakeJs/src/Main.purs:278]: I think that Nate showed this on the slack recently:

```purescript
import Control.MonadZero (guard)

fromBool x b = guard b $> x
```

**Comment by JordanMartinez** [comment]
> I give permission to use the example in the PureScript Cookbook.

@holdenlee Thanks, but under what terms? Per our [guidelines here](https://github.com/JordanMartinez/purescript-cookbook/blob/master/CONTRIBUTING.md#all-recipes-are-licensed-under-this-repos-license-mit), all recipes are licensed under MIT. Are you ok with that?

**Comment by holdenlee** [comment]
@JordanMartinez Yes, under the MIT license.

**Pull request merged**